x-pack/plugin/logsdb/src/yamlRestTest/resources/rest-api-spec/test/40_source_mode_setting.yml

```diff
@@ -0,0 +1,602 @@
+
+---
+create an index with time_series index mode and stored source:
+  - do:
+      catch: bad_request
+      indices.create:
+        index: test_time_series_index_mode_undefined
+        body:
+          settings:
+            index:
+              mode: time_series
+              mapping.source.mode: stored
+              routing_path: [ keyword ]
+              time_series:
+                start_time: 2021-04-28T00:00:00Z
+                end_time: 2021-04-29T00:00:00Z
+          mappings:
+            properties:
+              keyword:
+                type: keyword
+                time_series_dimension: true
+
+  - match: { error.type: "mapper_parsing_exception" }
+  - match: { error.reason: "Failed to parse mapping: time series indices only support synthetic source" }
+
+---
+create an index with logsdb index mode and stored source:
+  - do:
+      catch: bad_request
+      indices.create:
+        index: test_logsdb_index_mode_undefined
+        body:
+          settings:
+            index:
+              mode: logsdb
+              mapping.source.mode: stored
+
+  - match: { error.type: "mapper_parsing_exception" }
+  - match: { error.reason: "Failed to parse mapping: Indices with with index mode [logsdb] only support synthetic source" }
+
+---
+create an index with time_series index mode and disabled source:
+  - do:
+      catch: bad_request
+      indices.create:
+        index: test_time_series_index_mode
+        body:
+          settings:
+            index:
+              mode: time_series
+              mapping.source.mode: disabled
+              routing_path: [ keyword ]
+              time_series:
+                start_time: 2021-04-28T00:00:00Z
+                end_time: 2021-04-29T00:00:00Z
+          mappings:
+            properties:
+              keyword:
+                type: keyword
+                time_series_dimension: true
+
+  - match: { error.type: "mapper_parsing_exception" }
+  - match: { error.reason: "Failed to parse mapping: time series indices only support synthetic source" }
+
+---
+create an index with logsdb index mode and disabled source:
+  - do:
+      catch: bad_request
+      indices.create:
+        index: test_logsdb_index_mode
+        body:
+          settings:
+            index:
+              mode: logsdb
+              mapping.source.mode: disabled
+
+  - match: { error.type: "mapper_parsing_exception" }
+  - match: { error.reason: "Failed to parse mapping: Indices with with index mode [logsdb] only support synthetic source" }
+
+---
+modify final setting after index creation:
+  - do:
+      indices.create:
+        index: test_modify_setting
+        body:
+          settings:
+            index.mapping.source.mode: stored
+
+  - do:
+      catch: /.*Can't update non dynamic setting.*/
+      indices.put_settings:
+        index: test_modify_setting
+        body:
+          index:
+            mapping.source.mode: synthetic
+
+---
+modify source mapping after index creation:
```

Comment:
Note that this test fails before we even create the index in `MappingParser`
```
if (mappingParserContext.indexVersionCreated().isLegacyIndexVersion() == false) {
            // legacy indices are allowed to have extra definitions that we ignore (we will drop them on import)
            checkNoRemainingFields(mappingSource, "Root mapping definition has unsupported parameters: ");
        }
```

when parsing the put mapping request the `_source` mapping remains in the mappings not applied.

Reply:
This is because when you do `put_mapping` you don't need top level `mappings` entry.


```
  - do:
      indices.put_mapping:
        index: test_modify_source_mode
        body:
          _source:
            mode: synthetic
```

Reply:
I fixed that and added a few more tests to make sure that the setting always takes precedence over the source mode aven after put mapping is used on _source.mode